### PR TITLE
Add SpanStringTokenizer and avoid many string allocations in FontCollectionBase

### PR DESCRIPTION
**Pull request by Washi1337**
## What does the pull request do?

This PR creates a `ref struct` variant of `StringTokenizer` called `SpanStringTokenizer`, which holds a `ReadOnlySpan<char>` instead of a `System.String`. Additionally, this also adds `EnumHelpers::TryParse` shims accepting spans as input parameter. This drastically reduces the number of allocations of unnecessary `System.String` objects in `FontCollectionBase` and friends.

## What is the current behavior?

When creating a `TextRunProperties` (e.g., a `GenericTextRunProperties`), somewhere down the line Avalonia.Skia accesses `Typeface.GlyphTypeface`. This in turn eventually bubbles down all the way to `FontCollectionBase::GetImplicitTypeface`, which tries to normalize the typeface by carving out the font style, weight and stretch out of the full family name. It does so with the help of `StringTokenizer`.

For example, below is the logic for carving out the font weight out of the full family name:

https://github.com/AvaloniaUI/Avalonia/blob/44a784a8af2a4a9073e0979eaa27c1c7cdc00649/src/Avalonia.Base/Media/Fonts/FontCollectionBase.cs#L294-L320

`StringTokenizer` takes in a `System.String` object, meaning that all parsing needs to have a fully concretized string. This requires every single token returned by the tokenizer to be fully concretized to a `System.String`, or else we cannot determine whether it was an integer, nor can we pass it into `Enum.TryParse` to match it with font modifiers. Additionally, even more new string allocations are made by concatenating strings and calling `Replace` and `TrimEnd`. Finally, since there are three types of modifiers, there are three very similar parsing functions that all do extensive string object allocations. This all to simply parse a font family name.

image

While the glyph typeface is cached in the internal `CachedGlyphTypeface` property, only slightly adjusting the visual appearance of text requires a new `TextRunProperties`, and thus results in the creation of a new cached glyph typeface. This happens in projects such as [AvaloniaEdit](https://github.com/AvaloniaUI/AvaloniaEdit) and [AvaloniaHex](https://github.com/Washi1337/AvaloniaHex), which need to create many `TextRunProperties` instances to implement syntax highlighting. Additionally, every [TextBlock](https://github.com/AvaloniaUI/Avalonia/blob/44a784a8af2a4a9073e0979eaa27c1c7cdc00649/src/Avalonia.Controls/TextBlock.cs#L656) control creates a new instance as well. As far as I know, this property is not exposed nor can a `TextRunProperties` be cloned efficiently with this cached glyph typeface either. Even though, out of scope for this PR, this potentially could also be an interesting avenue for further optimizations.

## What is the updated/expected behavior with this PR?

Drastically reduced observed `System.String` allocations when creating many `TextRunProperties`:

image2

## How was the solution implemented (if it's not obvious)?

- A new `internal ref struct SpanStringTokenizer` which is an exact copy of `StringTokenizer` but stores a `ReadOnlySpan<char>` as opposed to a `string`.
- References to `StringTokenizer` were updated to `SpanStringTokenizer` throughout the entire Avalonia solution.
- `StringTokenizer` is now marked obsolete.
- `EnumHelpers` now include `TryParse` shims.
- Carving out the modifiers is now done via a lazy initialized `StringBuilder`. 

## Checklist

- [x] Added unit tests (if possible)?

Added a copy of `StringTokenizerTests` but specifically for `SpanStringTokenizer`.

## Breaking changes

The only technically publicly visible breaking change is turning `GridLength::ParseLenghts` into eagerly materializing a `List<GridLength>`, as `ref struct`s are not allowed in conjunction with `yield return`. However I don't think this should be much of an issue:

```diff
         public static IEnumerable<GridLength> ParseLengths(string s)
         {
-            using (var tokenizer = new StringTokenizer(s, CultureInfo.InvariantCulture))
+            var result = new List<GridLength>();
+
+            using (var tokenizer = new SpanStringTokenizer(s, CultureInfo.InvariantCulture))
             {
                 while (tokenizer.TryReadString(out var item))
                 {
-                    yield return Parse(item);
+                    result.Add(Parse(item));
                 }
             }
+
+            return result;
```

## Issues

- Related: #16390

### CODE REVIEW

**Comment by Washi1337** [comment]
Tests seem only to be failing because of the removal of `record` and `IDisposable` of the internal struct.

File: src/Avalonia.Base/Utilities/StringTokenizer.cs
Comment (kekekeks) [inline src/Avalonia.Base/Utilities/StringTokenizer.cs:11]: I believe that's a breaking change, so we can't have it until 12.0.

Not sure why is this class public in the first place.

We can probably a separate internal struct though.

Comment (Washi1337) [reply src/Avalonia.Base/Utilities/StringTokenizer.cs:11]: I could add an `internal ref struct SpanStringTokenizer` which will effectively be just a copy of the normal `StringTokenizer` but with the `ReadOnlySpan<char>` changes. We could then update all references in Avalonia itself while also keeping `StringTokenizer` as it is right now until we version bump to 12.0. Would that be an acceptable solution?

Comment (kekekeks) [reply src/Avalonia.Base/Utilities/StringTokenizer.cs:11]: I think so, yes. Also mark the current one as [Obsolete].

**Comment by Washi1337** [comment]
I reverted `StringTokenizer` and added `SpanStringTokenizer`. Not sure what happened to the macOS runner, but all other tests seem to pass now.

File: src/Avalonia.Base/Utilities/StringTokenizer.cs
Comment (timunie) [reply src/Avalonia.Base/Utilities/StringTokenizer.cs]: We should also add a comment  "TODO12" and probably add it also here https://github.com/AvaloniaUI/Avalonia/issues/12276

This in order to increase the chance it doesn't get forgotten.

File: src/Avalonia.Base/Media/Fonts/FontCollectionBase.cs
Comment (MrJul) [inline src/Avalonia.Base/Media/Fonts/FontCollectionBase.cs:373]: Why not use `Enum.TryParse<T>(ReadOnlySpan<char>, ...)` here? It would avoid maintaining the mappings manually.

A shim could be added to the existing `EnumHelper` class here for .NET Standard 2.0, falling back to the `string` overload in this case.

**Review by MrJul — COMMENTED** [review]
A remark aside, LGTM, thank you!
It's always good to have fewer allocations :)

**Comment by MrJul** [comment]
I'm sorry to hijack this PR for a moment.

@Alexizx00789999 could you please stop approving all pending pull requests? It won't help them get team approval any faster. This just adds unnecessary noise.

File: src/Avalonia.Base/Media/Fonts/FontCollectionBase.cs
Comment (Washi1337) [reply src/Avalonia.Base/Media/Fonts/FontCollectionBase.cs:373]: Ah I wasn't aware of the `EnumHelper`. I agree, this would be better from a maintainability perspective.

If we are fine with the netstandard2.0 build target to do more allocations than the newer .NET targets, then I will add the shims.

Comment (MrJul) [reply src/Avalonia.Base/Media/Fonts/FontCollectionBase.cs:373]: Yes, that's usually fine for .NET Standard. All `SpanHelpers` methods used by the `SpanStringTokenizer` are already falling back to the `string` overloads on that platform, effectively allocating.

If people want the best performance, they should target .NET 8.0+.

**Review by MrJul — APPROVED** [review]
LGTM, thank you!

**Pull request merged**
